libexec/rtld-elf/rtld_c18n.c

```diff
@@ -0,0 +1,700 @@
+{
+	string_handle i = 0;
+	do {
+		const char *cur = str;
+		while (sb->buf[i] == *cur++)
+			if (sb->buf[i++] == '\0')
+				return (i);
+		while (sb->buf[i] != '\0') ++i;
+	} while (++i < sb->size);
+	return (-1);
+}
+
+static char trusted_globals_names[] =
+	"memset\0"
+	"memcpy\0"
+	"mempcpy\0"
+	"memccpy\0"
+	"memchr\0"
+	"memrchr\0"
+	"memmove\0"
+	"strcpy\0"
+	"strncpy\0"
+	"stpcpy\0"
+	"stpncpy\0"
+	"strcat\0"
+	"strncat\0"
+	"strlcpy\0"
+	"strlcat\0"
+	"strlen\0"
+	"strnlen\0"
+	"strcmp\0"
+	"strncmp\0"
+
+	"__libc_start1\0"
+	"setjmp\0"
+	"longjmp\0"
+	"_setjmp\0"
+	"_longjmp\0"
+	"sigsetjmp\0"
+	"siglongjmp\0"
+
+	"vfork\0"
+	"rfork\0"
+
+	/* See comment in function definition */
+	"_rtld_thread_start";
+
+static const struct string_base trusted_globals = {
+	.buf = trusted_globals_names,
+	.size = sizeof(trusted_globals_names),
+	.capacity = sizeof(trusted_globals_names)
+};
+
+static struct compart rtld_compart = {
+	.name = _BASENAME_RTLD
+};
+
+static struct {
+	const struct compart **data;
+	compart_id_t size, capacity;
+} comparts;
+
+static void
+comparts_data_expand(compart_id_t capacity)
+{
+	const struct compart **data;
+
+	data = realloc(comparts.data, sizeof(*data) * capacity);
+	if (data == NULL)
+		rtld_fatal("realloc failed");
+	comparts.data = data;
+	comparts.capacity = capacity;
+}
+
+static bool
+string_search(const char *const strs[], const char *sym)
+{
+	if (strs != NULL)
+		for (; *strs != NULL; ++strs)
+			if (strcmp(sym, *strs) == 0)
+				return (true);
+	return (false);
+}
+
+static const struct compart *const *
+compart_get_or_create(const char *lib)
+{
+	struct {
+		const char *libraries[2];
+		struct compart com;
+	} *buf;
+	for (compart_id_t i = 1; i < comparts.size; ++i)
+		if (string_search(comparts.data[i]->libraries, lib))
+			return (&comparts.data[i]);
+	if (comparts.size == comparts.capacity)
+		comparts_data_expand(comparts.capacity * 2);
+	buf = xmalloc(sizeof(*buf));
+	buf->libraries[0] = lib;
+	buf->libraries[1] = NULL;
+	buf->com = (struct compart) {
+		.name = lib,
+		.libraries = buf->libraries
+	};
+	comparts.data[comparts.size] = &buf->com;
+	return (&comparts.data[comparts.size++]);
+}
+
+compart_id_t
+compart_id_allocate(const char *name)
+{
+	const struct compart *const *com = compart_get_or_create(name);
+	return (com - comparts.data);
+}
+
+static bool
+tramp_should_include(const Obj_Entry *reqobj, const struct tramp_data *data)
+{
+	const char *sym;
+
+	if (data->target == NULL)
+		return (false);
+
+	if (reqobj == NULL)
+		return (true);
+
+	if (reqobj->compart_id == data->defobj->compart_id)
+		return (false);
+
+	sym = strtab_value(data->defobj, data->def->st_name);
+
+	if (string_base_search(&trusted_globals, sym) != -1)
+		return (false);
+
+	return (true);
+}
+
+/*
+ * Trampolines
+ */
+struct tramp_pg {
+	void *_Atomic cursor;		/* Start of unused space */
+	SLIST_ENTRY(tramp_pg) link;	/* Link to next page */
+	void *trampolines[];		/* Start of trampolines */
```

Comment:
`void *` is a strange choice for code

Reply:
Indeed something like `uint32_t` would sound more appropriate. But because trampolines actually begin with data (capabilities), they have to be `void *` aligned.